Sources/GoogleCloudProvider/Storage/StorageRequest.swift

```diff
@@ -18,48 +18,68 @@ extension HTTPHeaders {
 
-        if let token = currentToken, let created = tokenCreatedTime, refreshableToken.isFresh(token: token, created: created) {
-            return try _send(method: method, headers: headers, path: path, query: query, body: body, accessToken: token.accessToken)
-        } else {
-            return try refreshableToken.refresh().flatMap({ (newToken) in
+    func send(method: HTTPMethod, headers: HTTPHeaders = [:], path: String, query: String, body: HTTPBody = HTTPBody()) throws -> Future<Data> {
+        return try withToken({ token in
+            return try self.send(method: method, headers: headers, path: path, query: query, body: body, accessToken: token.accessToken).flatMap({ response in
+                return response.http.body.consumeData(on: response)
+            })
+        })
+    }
+
+    private func withToken<F>(_ closure: @escaping (OAuthAccessToken) throws -> Future<F>) throws -> Future<F>{
+        guard let token = currentToken, let created = tokenCreatedTime, refreshableToken.isFresh(token: token, created: created) else {
+            return try refreshableToken.refresh().flatMap({ newToken in
                 self.currentToken = newToken
                 self.tokenCreatedTime = Date()
-                return try self._send(method: method, headers: headers, path: path, query: query, body: body, accessToken: newToken.accessToken)
+
+                return try closure(newToken)
             })
         }
+
+        return try closure(token)
     }
-    
-    private func _send<GCM: GoogleCloudModel>(method: HTTPMethod, headers: HTTPHeaders, path: String, query: String, body: HTTPBody, accessToken: String) throws -> Future<GCM> {
+
+    private func send(method: HTTPMethod, headers: HTTPHeaders, path: String, query: String, body: HTTPBody, accessToken: String) throws -> Future<Response> {
         var finalHeaders: HTTPHeaders = HTTPHeaders.gcsDefault
         finalHeaders.add(name: .authorization, value: "Bearer \(accessToken)")
         headers.forEach { finalHeaders.replaceOrAdd(name: $0.name, value: $0.value) }
-        
-        return httpClient.send(method, headers: finalHeaders, to: "\(path)?\(query)", beforeSend: { $0.http.body = body }).flatMap({ (response)  in
-            let decoder = JSONDecoder()
-            let formatter = DateFormatter()
-            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZ"
-            decoder.dateDecodingStrategy = .formatted(formatter)
+
+        return httpClient.send(method, headers: finalHeaders, to: "\(path)?\(query)", beforeSend: { $0.http.body = body }).flatMap({ response in
             guard response.http.status == .ok else {
-                return try decoder.decode(CloudStorageError.self, from: response.http, maxSize: 65_536, on: self.httpClient.container).map(to: GCM.self){ error in
+                return try self.responseDecoder.decode(CloudStorageError.self, from: response.http, maxSize: 65_536, on: self.httpClient.container).map(to: Response.self) { error in
```

Comment:
We also don't need to map this to a `Response` type if it's a CloudStorageError

Reply:
Done - we actually don't ever really return here, we intend to always throw an exception.